File: python/tests/test_mc.py
"""Monte Carlo: agreement with BS/PDE, antithetic effect, barrier properties."""

from __future__ import annotations

import math

import numpy as np
import pytest

from localvol import (
    Market,
    bs_price,
    price_european_mc,
    price_european_pde,
    price_up_out_call_mc,
)

SEED = 987


def test_flat_mc_matches_bs_within_3se():
    mkt = Market(100.0, 0.03, 0.01)
    exact = bs_price(100, 105, 0.03, 0.01, 0.2, 1.0, True)
    res = price_european_mc(mkt, 105.0, 1.0, 0.2, n_paths=20000, n_steps=50, seed=SEED)
    assert res.stderr > 0.0
    assert res.within(exact, 3.0), f"{res.price}±{res.stderr} vs {exact}"


def test_flat_mc_put_and_fx():
    mkt = Market.fx(1.20, rd=0.03, rf=0.01)
    exact = bs_price(1.20, 1.25, 0.03, 0.01, 0.10, 0.5, False)
    res = price_european_mc(mkt, 1.25, 0.5, 0.10, is_call=False, n_paths=20000, n_steps=50, seed=SEED)
    assert res.within(exact, 3.0)


def test_mc_deterministic_given_seed():
    mkt = Market(100.0, 0.02, 0.0)
    a = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=2000, n_steps=20, seed=11)
    b = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=2000, n_steps=20, seed=11)
    c = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=2000, n_steps=20, seed=12)
    assert a.price == b.price and a.stderr == b.stderr
    assert a.price != c.price


def test_antithetic_reduces_stderr():
    mkt = Market(100.0, 0.02, 0.0)
    anti = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=20000, n_steps=25, seed=SEED, antithetic=True)
    plain = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=20000, n_steps=25, seed=SEED, antithetic=False)
    assert anti.stderr < plain.stderr


def test_localvol_mc_matches_pde_within_3se(bundled_localvol):
    mkt = Market(100.0, 0.0, 0.0)
    pde = price_european_pde(
        mkt, 100.0, 1.0, bundled_localvol.vol, num_space=200, num_time=200, sigma_ref=0.2
    )
    res = price_european_mc(mkt, 100.0, 1.0, bundled_localvol.vol, n_paths=20000, n_steps=100, seed=SEED)
    assert res.within(pde, 3.0), f"MC {res.price}±{res.stderr} vs PDE {pde}"


def test_barrier_below_vanilla_and_bridge_lowers_price(bundled_localvol):
    mkt = Market(100.0, 0.02, 0.0)
    kwargs = dict(n_paths=10000, n_steps=50, seed=SEED)
    vanilla = price_european_mc(mkt, 100.0, 1.0, bundled_localvol.vol, **kwargs)
    plain = price_up_out_call_mc(mkt, 100.0, 125.0, 1.0, bundled_localvol.vol, brownian_bridge=False, **kwargs)
    bridged = price_up_out_call_mc(mkt, 100.0, 125.0, 1.0, bundled_localvol.vol, brownian_bridge=True, **kwargs)
    assert plain.price <= vanilla.price + 1e-12          # knock-out cannot add value
    assert bridged.price <= plain.price + 1e-12          # bridge sees more crossings
    assert bridged.price > 0.0


def test_barrier_born_knocked_out():
    mkt = Market(130.0, 0.02, 0.0)
    res = price_up_out_call_mc(mkt, 100.0, 125.0, 1.0, 0.2, n_paths=1000, n_steps=10, seed=1)
    assert res.price == 0.0 and res.stderr == 0.0


def test_validation():
    mkt = Market(100.0, 0.02, 0.0)
    with pytest.raises(ValueError, match="even"):
        price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=1001, antithetic=True)
    with pytest.raises(ValueError, match="n_steps"):
        price_european_mc(mkt, 100.0, 1.0, 0.2, n_steps=0)
    with pytest.raises(ValueError, match="expiry"):
        price_european_mc(mkt, 100.0, 0.0, 0.2)
    with pytest.raises(ValueError, match="flat vol"):
        price_european_mc(mkt, 100.0, 1.0, 0.0)
    with pytest.raises(ValueError, match="barrier"):
        price_up_out_call_mc(mkt, 100.0, -5.0, 1.0, 0.2)
    with pytest.raises(ValueError, match="strike"):
        price_european_mc(mkt, -1.0, 1.0, 0.2)


@pytest.mark.parametrize(
    "mkt",
    [Market(100.0, 0.05, 0.02), Market.fx(1.10, rd=0.03, rf=-0.01)],
    ids=["equity_r5_q2", "fx_rd3_rf_minus1"],
)
def test_localvol_mc_matches_pde_with_carry(bundled_localvol, mkt):
    """MAJ-7 regression: with r != q the MC lookup k = X_n - ln F(t_n) is
    actually exercised (a ln S0 lookup lands 4-5 SE away from the PDE)."""
    srf = bundled_localvol.surface
    strike, expiry = mkt.spot, 1.0
    pde = price_european_pde(
        mkt, strike, expiry, bundled_localvol.vol, num_space=200, num_time=200,
        sigma_ref=float(srf.implied_vol(0.0, expiry)),
    )
    res = price_european_mc(mkt, strike, expiry, bundled_localvol.vol, n_paths=20000, n_steps=100, seed=SEED)
    assert res.within(pde, 3.0), f"MC {res.price}±{res.stderr} vs PDE {pde}"


def test_mc_put_call_parity_under_local_vol(bundled_localvol):
    mkt = Market(100.0, 0.03, 0.01)
    kwargs = dict(n_paths=20000, n_steps=100, seed=5)
    c = price_european_mc(mkt, 100.0, 1.0, bundled_localvol.vol, is_call=True, **kwargs)
    p = price_european_mc(mkt, 100.0, 1.0, bundled_localvol.vol, is_call=False, **kwargs)
    parity = 100.0 * math.exp(-0.01) - 100.0 * math.exp(-0.03)
    assert abs(c.price - p.price - parity) < 3.0 * math.sqrt(c.stderr**2 + p.stderr**2)


def test_mc_rejects_tiny_path_counts():
    """MAJ-8 regression: n_paths = 2 with antithetic (one pair mean) used to
    return stderr = NaN; now rejected.  n_paths = 4 gives a finite stderr."""
    mkt = Market(100.0, 0.02, 0.0)
    with pytest.raises(ValueError, match="n_paths"):
        price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=2, n_steps=5, antithetic=True)
    with pytest.raises(ValueError, match="n_paths"):
        price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=1, n_steps=5, antithetic=False)
    with pytest.raises(ValueError, match="n_paths"):
        price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, 0.2, n_paths=2, n_steps=5, antithetic=True)
    # deep ITM strike so every path pays and the sample spread is non-zero
    res = price_european_mc(mkt, 50.0, 1.0, 0.2, n_paths=4, n_steps=5, antithetic=True)
    assert math.isfinite(res.price) and math.isfinite(res.stderr) and res.stderr > 0.0
    res = price_european_mc(mkt, 50.0, 1.0, 0.2, n_paths=2, n_steps=5, antithetic=False)
    assert math.isfinite(res.stderr) and res.stderr > 0.0


def test_mc_rejects_nonfinite_or_negative_vol_callable():
    """MAJ-4 regression: a bad callable used to yield price = stderr = NaN."""
    mkt = Market(100.0, 0.02, 0.0)
    with pytest.raises(ValueError, match="non-finite/negative"):
        price_european_mc(mkt, 100.0, 1.0, lambda k, t: np.nan * k, n_paths=100, n_steps=5)
    with pytest.raises(ValueError, match="non-finite/negative"):
        price_european_mc(mkt, 100.0, 1.0, lambda k, t: np.full_like(k, -0.2), n_paths=100, n_steps=5)
    with pytest.raises(ValueError, match="non-finite/negative"):
        price_european_mc(mkt, 100.0, 1.0, lambda k, t: np.full_like(k, np.inf), n_paths=100, n_steps=5)
    with pytest.raises(ValueError, match="non-finite/negative"):
        price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, lambda k, t: np.nan * k, n_paths=100, n_steps=5)
    with pytest.raises(ValueError, match="shape"):
        price_european_mc(mkt, 100.0, 1.0, lambda k, t: np.zeros(3), n_paths=100, n_steps=5)
    # A scalar-returning callable is broadcast and behaves like the flat vol.
    flat = price_european_mc(mkt, 100.0, 1.0, 0.2, n_paths=2000, n_steps=10, seed=3)
    scal = price_european_mc(mkt, 100.0, 1.0, lambda k, t: 0.2, n_paths=2000, n_steps=10, seed=3)
    assert scal.price == flat.price and scal.stderr == flat.stderr


def test_mc_zero_vol_callable_is_deterministic_forward():
    """sigma == 0 from a callable is allowed: every path is the forward."""
    mkt = Market(100.0, 0.03, 0.01)
    res = price_european_mc(mkt, 90.0, 1.0, lambda k, t: np.zeros_like(k), n_paths=8, n_steps=4, seed=1)
    assert res.price == pytest.approx(math.exp(-0.03) * (100.0 * math.exp(0.02) - 90.0), abs=1e-12)
    assert res.stderr == 0.0
    uo = price_up_out_call_mc(mkt, 90.0, 130.0, 1.0, lambda k, t: np.zeros_like(k), n_paths=8, n_steps=4, seed=1)
    assert uo.price == pytest.approx(res.price, abs=1e-12)  # never crosses: no bridge weight

File: cpp/tests/test_mc.cpp
// Monte Carlo: BS agreement within SE bands, antithetic constraints,
// barrier ordering (bridge <= discrete <= vanilla), validation.

#include <gtest/gtest.h>

#include <cmath>
#include <limits>
#include <stdexcept>
#include <string>

#include "localvol/black_scholes.hpp"
#include "localvol/dupire.hpp"
#include "localvol/market.hpp"
#include "localvol/mc.hpp"
#include "localvol/pde.hpp"
#include "localvol/surface.hpp"

using localvol::bs_price;
using localvol::DupireLocalVol;
using localvol::ImpliedVolSurface;
using localvol::Market;
using localvol::McResult;
using localvol::McSettings;
using localvol::PdeSettings;
using localvol::price_european_mc;
using localvol::price_european_pde;
using localvol::price_up_out_call_mc;
using localvol::VolFn;

namespace {
const std::string kDataDir = LOCALVOL_DATA_DIR;

McSettings settings(int n_paths, int n_steps, std::uint64_t seed = 42) {
    McSettings s;
    s.n_paths = n_paths;
    s.n_steps = n_steps;
    s.seed = seed;
    return s;
}
}  // namespace

TEST(Mc, FlatVolMatchesBlackScholesWithin3Se) {
    const Market mkt(100.0, 0.03, 0.01);
    const double exact = bs_price(100, 105, 0.03, 0.01, 0.2, 1.0, true);
    const McResult res = price_european_mc(mkt, 105.0, 1.0, 0.2, true, settings(20000, 50));
    EXPECT_GT(res.std_err, 0.0);
    EXPECT_TRUE(res.within(exact, 3.0))
        << "MC " << res.price << " +/- " << res.std_err << " vs BS " << exact;
}

TEST(Mc, PutFlavourAndDeterminism) {
    const Market mkt(100.0, 0.02, 0.0);
    const double exact = bs_price(100, 95, 0.02, 0.0, 0.25, 0.5, false);
    const McResult a = price_european_mc(mkt, 95.0, 0.5, 0.25, false, settings(10000, 50));
    const McResult b = price_european_mc(mkt, 95.0, 0.5, 0.25, false, settings(10000, 50));
    EXPECT_TRUE(a.within(exact, 4.0));
    EXPECT_DOUBLE_EQ(a.price, b.price);  // fixed seed -> deterministic
    EXPECT_DOUBLE_EQ(a.std_err, b.std_err);
}

TEST(Mc, AntitheticReducesStandardError) {
    const Market mkt(100.0, 0.03, 0.0);
    McSettings anti = settings(20000, 25);
    McSettings plain = settings(20000, 25);
    plain.antithetic = false;
    const McResult ra = price_european_mc(mkt, 100.0, 1.0, 0.2, true, anti);
    const McResult rp = price_european_mc(mkt, 100.0, 1.0, 0.2, true, plain);
    EXPECT_LT(ra.std_err, rp.std_err);  // ATM payoff: pairing helps
}

TEST(Mc, CallableVolMatchesFlatVol) {
    const Market mkt(100.0, 0.01, 0.0);
    const McResult flat = price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(4000, 20));
    const McResult fn = price_european_mc(
        mkt, 100.0, 1.0, [](double, double) { return 0.2; }, true, settings(4000, 20));
    EXPECT_DOUBLE_EQ(flat.price, fn.price);  // identical draws, identical sigma
}

TEST(Mc, BarrierOrderingBridgeDiscreteVanilla) {
    const Market mkt(100.0, 0.02, 0.0);
    const McSettings s = settings(20000, 100);
    const McResult vanilla = price_european_mc(mkt, 100.0, 1.0, 0.2, true, s);
    const McResult discrete = price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, 0.2, s, false);
    const McResult bridged = price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, 0.2, s, true);
    EXPECT_LE(bridged.price, discrete.price);  // bridge removes upward bias
    EXPECT_LE(discrete.price, vanilla.price);  // barrier <= vanilla
    EXPECT_GT(bridged.price, 0.0);
}

TEST(Mc, BarrierBornKnockedOutIsZero) {
    const Market mkt(140.0, 0.02, 0.0);
    const McResult res = price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, 0.2, settings(1000, 10));
    EXPECT_DOUBLE_EQ(res.price, 0.0);
    EXPECT_DOUBLE_EQ(res.std_err, 0.0);
}

TEST(Mc, RejectsInvalidInputs) {
    const Market mkt(100.0, 0.02, 0.0);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 0.0, 0.2, true, settings(1000, 10)),
                 std::invalid_argument);  // expiry must be > 0
    EXPECT_THROW(price_european_mc(mkt, -1.0, 1.0, 0.2, true, settings(1000, 10)),
                 std::invalid_argument);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.0, true, settings(1000, 10)),
                 std::invalid_argument);  // flat vol must be > 0 for MC
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(1001, 10)),
                 std::invalid_argument);  // antithetic needs even n_paths
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(1000, 0)),
                 std::invalid_argument);
    McSettings tiny = settings(1, 10);
    tiny.antithetic = false;
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.2, true, tiny), std::invalid_argument);
    EXPECT_THROW(price_up_out_call_mc(mkt, 100.0, -130.0, 1.0, 0.2, settings(1000, 10)),
                 std::invalid_argument);
}

TEST(Mc, LocalVolMatchesPdeWithCarry) {
    // MAJ-7: with r != q the lookup k = X_n - ln F(t_n) is exercised
    // (a ln S0 lookup lands 4-5 SE away from the PDE).
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/implied_surface.csv");
    const DupireLocalVol lv(srf);
    const VolFn vol = [&lv](double k, double t) { return lv.vol(k, t); };
    const Market markets[] = {Market(100.0, 0.05, 0.02), Market::fx(1.10, 0.03, -0.01)};
    for (const Market& mkt : markets) {
        const double strike = mkt.spot();
        const double pde =
            price_european_pde(mkt, strike, 1.0, vol, true, PdeSettings{}, srf.implied_vol(0.0, 1.0));
        const McResult res = price_european_mc(mkt, strike, 1.0, vol, true, settings(20000, 100, 987));
        EXPECT_TRUE(res.within(pde, 3.0))
            << "S0=" << mkt.spot() << ": MC " << res.price << " +/- " << res.std_err << " vs PDE " << pde;
    }
}

TEST(Mc, PutCallParityUnderLocalVol) {
    const ImpliedVolSurface srf = ImpliedVolSurface::from_csv(kDataDir + "/implied_surface.csv");
    const DupireLocalVol lv(srf);
    const VolFn vol = [&lv](double k, double t) { return lv.vol(k, t); };
    const Market mkt(100.0, 0.03, 0.01);
    const McSettings s = settings(20000, 100, 5);
    const McResult c = price_european_mc(mkt, 100.0, 1.0, vol, true, s);
    const McResult p = price_european_mc(mkt, 100.0, 1.0, vol, false, s);
    const double parity = 100.0 * std::exp(-0.01) - 100.0 * std::exp(-0.03);
    EXPECT_LT(std::abs(c.price - p.price - parity),
              3.0 * std::sqrt(c.std_err * c.std_err + p.std_err * p.std_err));
}

TEST(Mc, RejectsTinyPathCounts) {
    // MAJ-8: n_paths = 2 with antithetic used to return std_err = NaN.
    const Market mkt(100.0, 0.02, 0.0);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(2, 5)),
                 std::invalid_argument);
    McSettings one = settings(1, 5);
    one.antithetic = false;
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, 0.2, true, one), std::invalid_argument);
    EXPECT_THROW(price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, 0.2, settings(2, 5)),
                 std::invalid_argument);
    // deep ITM strike so every path pays and the sample spread is non-zero
    const McResult four = price_european_mc(mkt, 50.0, 1.0, 0.2, true, settings(4, 5));
    EXPECT_TRUE(std::isfinite(four.price));
    EXPECT_TRUE(std::isfinite(four.std_err));
    EXPECT_GT(four.std_err, 0.0);
    McSettings two = settings(2, 5);
    two.antithetic = false;
    const McResult r2 = price_european_mc(mkt, 50.0, 1.0, 0.2, true, two);
    EXPECT_TRUE(std::isfinite(r2.std_err));
    EXPECT_GT(r2.std_err, 0.0);
}

TEST(Mc, RejectsNonFiniteOrNegativeVolCallable) {
    // MAJ-4: a bad callable used to yield price = std_err = NaN.
    const Market mkt(100.0, 0.02, 0.0);
    const McSettings s = settings(100, 5);
    const VolFn nan_fn = [](double, double) { return std::nan(""); };
    const VolFn neg_fn = [](double, double) { return -0.2; };
    const VolFn inf_fn = [](double, double) { return std::numeric_limits<double>::infinity(); };
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, nan_fn, true, s), std::invalid_argument);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, neg_fn, true, s), std::invalid_argument);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, inf_fn, true, s), std::invalid_argument);
    EXPECT_THROW(price_up_out_call_mc(mkt, 100.0, 130.0, 1.0, nan_fn, s), std::invalid_argument);
    EXPECT_THROW(price_european_mc(mkt, 100.0, 1.0, VolFn{}, true, s), std::invalid_argument);
    // sigma == 0 from a callable is allowed: every path is the forward.
    const Market m2(100.0, 0.03, 0.01);
    const VolFn zero = [](double, double) { return 0.0; };
    const McResult res = price_european_mc(m2, 90.0, 1.0, zero, true, settings(8, 4));
    EXPECT_NEAR(res.price, std::exp(-0.03) * (100.0 * std::exp(0.02) - 90.0), 1e-12);
    EXPECT_DOUBLE_EQ(res.std_err, 0.0);
    const McResult uo = price_up_out_call_mc(m2, 90.0, 130.0, 1.0, zero, settings(8, 4), true);
    EXPECT_NEAR(uo.price, res.price, 1e-12);  // never crosses: no bridge weight
}

TEST(Mc, DifferentSeedsDiffer) {
    const Market mkt(100.0, 0.02, 0.0);
    const McResult a = price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(2000, 20, 11));
    const McResult b = price_european_mc(mkt, 100.0, 1.0, 0.2, true, settings(2000, 20, 12));
    EXPECT_NE(a.price, b.price);
}
